logging/google/cloud/logging/_shutdown.py

```diff
@@ -0,0 +1,131 @@
+from google.cloud.logging._environment_vars import _APPENGINE_FLEXIBLE_ENV_FLEX
+
+
+# Maximum size in bytes to send to Stackdriver Logging in one entry
+MAX_PAYLOAD_SIZE = 1024 * 100
+
+
+def _get_gae_instance():
+    """Returns the App Engine Flexible instance."""
+    return os.getenv('GAE_INSTANCE')
+
+
+def _get_gae_service():
+    """Returns the App Engine Flexible service."""
+    return os.getenv('GAE_SERVICE')
+
+
+def _get_gae_version():
+    """Returns the App Engine Flexible version."""
+    return os.getenv('GAE_VERSION')
+
+
+def _split_entry(payload):
+    """Splits payload into lists of maximum 100Kb.
+
+    Stackdriver Logging payloads are a maximum of 100Kb.
+    """
+    return [payload[i:i + MAX_PAYLOAD_SIZE]
+            for i in range(0, len(payload), MAX_PAYLOAD_SIZE)]
+
+
+def _write_stacktrace_log(client, traces):
+    """Writes the trace logs to the appropriate GAE resource in Stackdriver.
+
+    :type client: `google.cloud.logging.Client`
+    :param client: Stackdriver logging client.
+
+    :type traces: str
+    :param traces: String containing the stacktrace info to write to
+                   Stackdriver logging.
+    """
+    gae_version = _get_gae_version()
+    gae_service = _get_gae_service()
+    gae_instance = _get_gae_instance()
+
+    text_payload = '{}\nThread traces\n{}'.format(gae_instance, traces)
+
+    logger_name = 'projects/{}/logs/{}'.format(
+        client.project, 'appengine.googleapis.com%2Fapp.shutdown')
+
+    resource = {'type': 'gae_app', 'labels': {'project_id': client.project,
+                                              'version_id': gae_version,
+                                              'module_id': gae_service}}
+
+    labels = {
+        'appengine.googleapis.com/version_id': gae_version,
+        'compute.googleapis.com/resource_type': 'instance',
+        'appengine.googleapis.com/instance_name': gae_instance,
+        'appengine.googleapis.com/module_id': gae_service,
+    }
+
+    split_payloads = _split_entry(text_payload.encode('utf-8'))
+    entries = [{'text_payload': payload} for payload in split_payloads]
+
+    client.logging_api.write_entries(
+        entries, logger_name=logger_name, resource=resource, labels=labels)
+
+
+def _is_on_appengine():
+    """Returns True if the environment is detected as App Engine flexible."""
+    return (os.getenv(_APPENGINE_FLEXIBLE_ENV_VM) or os.getenv(
+        _APPENGINE_FLEXIBLE_ENV_FLEX))
+
+
+def _report_stacktraces(
+        client, signal_, frame):  # pylint: disable=unused-argument
+    """Reports the stacktraces of all active threads to Stackdriver Logging.
+
+    :type client: `google.cloud.logging.Client`
+    :param client: Stackdriver logging client.
+
+    :type signal_: int
+    :param signal_: Signal number.
+
+    :type frame: frame object
+    :param frame: The current stack frame.
+    """
+    traces = ''
+    for thread_id, stack in sys._current_frames().items():
```

Comment:
I totally agree but so far don't see a better option, virtually every search result is recommending this approach, and I think the whole concept of this PR is slightly exotic enough that maybe sys.py authors thought it wasn't necessary to make public. I can't find any obvious alternative...if it bothers you a lot maybe I could shoot out an email to various mailing lists?